lib/Parse/ParseExpr.cpp

```diff
@@ -2813,13 +2813,71 @@ ExprResult Parser::ParseBoundsExpression() {
     // The identifier is not a valid contextual keyword for the start of a
     // a bounds expression.
     Diag(Tok, diag::err_expected_bounds_expr);
+    SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
     Result = ExprError();
   }
 
   PT.consumeClose();
   return Result;
 }
 
+/// Consume and store tokens for an expression shaped like a bounds expression
+/// in the passed token container. Returns \c true if it reached the end of
+/// something bounds-expression shaped, \c false if a parsing error occurred,
+///
+/// Stop when the end of the expression is reached or a parsing error occurs
+/// for which ParseBoundsExpression doesn't know how to make forward progress.
+/// An expression is shaped like a bounds expression if it consists of
+///   identifier '(' sequence of tokens ')'
+/// where parentheses balance within the sequence of tokens.
+bool Parser::ConsumeAndStoreBoundsExpression(CachedTokens &Toks) {
+  Toks.push_back(Tok);
+  if (Tok.getKind() != tok::identifier) {
+    return false;
+  }
+  ConsumeToken();
+
+  Toks.push_back(Tok);
+  if (Tok.getKind() != tok::l_paren) {
+    return false;
+  }
+  ConsumeParen();
+
+  return ConsumeAndStoreUntil(tok::r_paren, Toks, /*StopAtSemi=*/true);
+}
+
+/// Given a list of tokens that have the same shape as a bounds
+/// expression, parse them to create a bounds expression.
+ExprResult Parser::DeferredParseBoundsExpression(CachedTokens &Toks) {
+  Token LastBoundsExprToken = Toks.back();
+  Token BoundsExprEnd;
+  BoundsExprEnd.startToken();
+  BoundsExprEnd.setKind(tok::eof);
+  SourceLocation OrigLoc = LastBoundsExprToken.getEndLoc();
+  BoundsExprEnd.setLocation(OrigLoc);
+  Toks.push_back(BoundsExprEnd);
```

Comment:
>push_back [](start = 7, length = 9)

It's a bit odd that this code is mutating its input.  At the risk of copying data and consuming slightly more memory, I think it would be cleaner to have the Toks parameter be const and use a separate private list here if you can...